### ISSUE TITLE
Satisfiability-by-Value in pattern matcher.

### ISSUE THREAD

**Issue by linas** [opencog/atomspace#1893]
Currently, the pattern matcher uses `PresentLink` and `AbsentLink`, together with `NotLink`, `AndLink`, `ChoiceLink, etc. to determine is a graph is groundable.  This makes some types of pattern searches awkward: for example, finding all graphs with an attention value greater than 42, or with a truth-value greater than (0.62, 0.44), or with some other generic value.

Currently, applying these additional constraints is ad-hoc: the user first finds all matching patterns, then writes some custom post-processing code, in scheme/python, to discard all atoms with the wrong attention value or truth value.  This has two or three downsides:

* If the post-processing discards 99% or all search results, then it would have been faster to perform the discard during search, rather than after.
* The post-processing code is ad-hoc, and not atomese. If it was atomese, then it could (potentially) be compiled to bytecode, for performance.
* Lack of tinkerpop/gremlin compatibility.

Lets focus on the last: a Gremlin-language traversal is just like a single-clause pattern match (Gremlin has no concept of a multi-clause traversal).  Except ... Gremlin traversals allow any access to values, and not just Present/Absent.  This suggests that the pattern matcher could be extended with new predicates to check values. This is already possible, but is, again, ad-hoc and poorly understood (most users don't realize this).  Examples provided below.

**Comment by linas** [comment]
Example: user wishes to check if the attention value of `Variable $X` is greater than 42 during traversal.  Then the `BindLink` or `GetLink` or `SatisfactionLink` would need to include a clause:
```
EvaluationLink
    GroundedPredicateNode "scm:is-av-sti-big-enough"
    ListLink
          Variable $X
           Number 42
```
with `is-av-big-enough` some small scheme snippet to examine the AV value, e.g. 
```
    (define (is-av-big-enough ato num)
          (if (< (float-from-string (cog-name num)) (cog-av-sti atom)) (stv 1 1) (stv 1 0))))
```
which is not hard to write, but still feels awkward.  Replacing the above with a single `STIGreaterThanLink` might be more elegant...

**Comment by linas** [comment]
Example: user wishes to find all atoms having the `StringValue` of `"Kevin"` under the key `"employee name"`.   Then the `BindLink` or `GetLink or `SatisfactionLink` would need to include a clause:

```
EvaluationLink
    GroundedPredicateNode "scm:key-has-string-value"
    ListLink
          Variable $X
          Predicate "employee name"
          Node "Kevin"
          NumberNode 0
```

with `key-has-string-value`  some small scheme snippet to examine the `StringValue` value, e.g.

    (define (key-has-string-value ato key val idx)
          (if (equal? (cog-name val) (cog-value-ref (cog-value ato key) (cog-number idx))) 
                (stv 1 1) (stv 1 0))))
```
which is not hard to write, but still feels awkward. Replacing the above with a single `StringValueEqualLink` might be more elegant...

**Comment by linas** [comment]
To be clear: all that this is really calling for is just an expansion of the set of pre-defined virtual links. We already have a `GreaterThanLink`, and a handful of others. No actual changes to the pattern matcher are required: the pattern matcher already supports these kinds of queries. This is actually just a request for more user-friendly link types, to make queries simpler to write (and simpler to read!?), and maybe enable a simpler way to have a tinkerpop/gremlin compat-like API.

**Comment by linas** [comment]
Closing, implemented in #2033 -- see `/examples/pattern-matcher/value-of.scm`
